examples/gateway/car-file-gateway/api.go

```diff
@@ -0,0 +1,221 @@
+	}
+
+	node, err := api.dsrv.Get(ctx, rp.Cid())
+	if err != nil {
+		return nil, fmt.Errorf("get node: %w", err)
+	}
+	return node, nil
+}
+
+func (api *carGateway) GetUnixFsNode(ctx context.Context, pth path.Path) (files.Node, error) {
+	nd, err := api.resolveNode(ctx, pth)
+	if err != nil {
+		return nil, err
+	}
+
+	return unixfile.NewUnixfsFile(ctx, api.dsrv, nd)
+}
+
+func (api *carGateway) LsUnixFsDir(ctx context.Context, p path.Path, opts ...options.UnixfsLsOption) (<-chan iface.DirEntry, error) {
+	settings, err := options.UnixfsLsOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	dagnode, err := api.resolveNode(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	dir, err := uio.NewDirectoryFromNode(api.dsrv, dagnode)
+	if err == uio.ErrNotADir {
+		return api.lsFromLinks(ctx, dagnode.Links(), settings)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return api.lsFromLinksAsync(ctx, dir, settings)
+}
+
+func (api *carGateway) processLink(ctx context.Context, linkres ft.LinkResult, settings *options.UnixfsLsSettings) iface.DirEntry {
+	if linkres.Err != nil {
+		return iface.DirEntry{Err: linkres.Err}
+	}
+
+	lnk := iface.DirEntry{
+		Name: linkres.Link.Name,
+		Cid:  linkres.Link.Cid,
+	}
+
+	switch lnk.Cid.Type() {
+	case cid.Raw:
+		// No need to check with raw leaves
+		lnk.Type = iface.TFile
+		lnk.Size = linkres.Link.Size
+	case cid.DagProtobuf:
+		if !settings.ResolveChildren {
+			break
+		}
+
+		linkNode, err := linkres.Link.GetNode(ctx, api.dsrv)
+		if err != nil {
+			lnk.Err = err
+			break
+		}
+
+		if pn, ok := linkNode.(*merkledag.ProtoNode); ok {
+			d, err := ft.FSNodeFromBytes(pn.Data())
+			if err != nil {
+				lnk.Err = err
+				break
+			}
+			switch d.Type() {
+			case ft.TFile, ft.TRaw:
+				lnk.Type = iface.TFile
+			case ft.THAMTShard, ft.TDirectory, ft.TMetadata:
+				lnk.Type = iface.TDirectory
+			case ft.TSymlink:
+				lnk.Type = iface.TSymlink
+				lnk.Target = string(d.Data())
+			}
+			lnk.Size = d.FileSize()
+		}
+	}
+
+	return lnk
+}
+
+func (api *carGateway) lsFromLinksAsync(ctx context.Context, dir uio.Directory, settings *options.UnixfsLsSettings) (<-chan iface.DirEntry, error) {
+	out := make(chan iface.DirEntry, uio.DefaultShardWidth)
+
+	go func() {
+		defer close(out)
+		for l := range dir.EnumLinksAsync(ctx) {
+			select {
+			case out <- api.processLink(ctx, l, settings): // TODO: perf: processing can be done in background and in parallel
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return out, nil
+}
+
+func (api *carGateway) lsFromLinks(ctx context.Context, ndlinks []*ipld.Link, settings *options.UnixfsLsSettings) (<-chan iface.DirEntry, error) {
+	links := make(chan iface.DirEntry, len(ndlinks))
+	for _, l := range ndlinks {
+		lr := ft.LinkResult{Link: &ipld.Link{Name: l.Name, Size: l.Size, Cid: l.Cid}}
+
+		links <- api.processLink(ctx, lr, settings) // TODO: can be parallel if settings.Async
+	}
+	close(links)
+	return links, nil
+}
```

Comment:
Code borrowed from https://github.com/iand/meridian. I wonder if there's an easier way to do this.